src/software/ai/hl/stp/stp.cpp

```diff
@@ -137,17 +145,44 @@ std::vector<std::shared_ptr<Tactic>> STP::assignRobotsToTactics(
     // https://github.com/saebyn/munkres-cpp is the implementation of the Hungarian
     // algorithm that we use here
 
-    if (world.friendlyTeam().numRobots() < tactics.size())
+    auto friendly_team         = world.friendlyTeam();
+    auto& friendly_team_robots = friendly_team.getAllRobots();
+
+    // Special handling for the Goalie tactics, since only one robot per team is permitted
+    // to act as the goalie
+    const std::optional<Robot> goalie    = friendly_team.goalie();
+    std::vector<Robot> non_goalie_robots = friendly_team_robots;
+    auto isGoalieTactic                  = [](std::shared_ptr<Tactic> tactic) {
+        return tactic->isGoalieTactic();
+    };
+
+    auto it = tactics.begin();
+    if (goalie)
+    {
+        non_goalie_robots.erase(
+            std::find(non_goalie_robots.begin(), non_goalie_robots.end(), *goalie));
+
+        // Assign the goalie to the first goalie tactic
+        auto iter = std::find_if(tactics.begin(), tactics.end(), isGoalieTactic);
+        if (iter != tactics.end())
+        {
+            (*iter)->updateRobot(*goalie);
+        }
+    }
+
+    // Discard all goalie tactics, since we have already assigned the goalie robot (if
+    // there is one) to the first goalie tactic, and there should only ever be one goalie
+    tactics.erase(std::remove_if(tactics.begin(), tactics.end(), isGoalieTactic),
```

Comment:
Why are we erasing the return value of remove_if?

Reply:
This is the [remove-erase idiom](https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom), it's how you're supposed to use `std::remove_if`, and saves on memory moving.

Reply:
Ah I see. This is necessary for ordered data structures. Previously, I'd been using sets